release/run_release_test.sh

```diff
@@ -66,6 +69,18 @@ if [ -z "${NO_CLONE}" ]; then
   TMPDIR=$(mktemp -d -t release-XXXXXXXXXX)
   git clone --depth 1 -b "${RAY_TEST_BRANCH}" "${RAY_TEST_REPO}" "${TMPDIR}"
   pushd "${TMPDIR}/release" || true
+  HEAD_COMMIT=$(git rev-parse HEAD)
+
+  # We only do this if RAY_TEST_REPO and RAY_TEST_BRANCH are pointing to ray master.
```

Comment:
Shouldn't we do this always?

Reply:
Release tests can be launched either automatically (I) or through PR (II). 

For II, there are times that these two commits are intended to be different. On one hand, I want to use my own repository and my own PR commit for testing code. On the other hand, the wheel is still the latest wheel on master. This is useful when people update some test code and does not change source code. 

For the racing condition we are trying to address, it mostly applies to II.

For II, sequence of things that happen:
0. Latest wheel on master is wheel_n
1. a PR (that modifies both the test code and the source code) is landed.
2. a wheel is being built (let's call it wheel_n+1)
3. Now nightly release test starts to run.
4. in `run_release_test.sh`, we fetch most recent wheel and install it. This is wheel_n, since wheel_n+1 is still being built.
5. in `run_release_tst.sh`, we check out test code, which is current master (includes PR_n+1).
6. Now there is a discrepancy between source code (wheel version) and test code.

It seems to me that the way to resolve this (most reliably) is to deduce commit hash directly from the wheel installed. 

For the clone step in `build_pipeline.py`, are you referring to 
```
clone_cmd = f"git clone --depth 1 --branch {branch} {repo} {tmpdir}"
```
?

Let's say the branch and repo is ray master. 
This clone may still be subject to inconsistency right? No one guarantees that this clone will have the same head commit hash as the ray_wheel right? So it shouldn't be used as the source of truth.

Reply:
I think with this solution a problem comes up when you specify a wheel commit to test against (e.g. for bisecting). E.g. to test the current version of a release test (on the Ray master repo) against past wheels. This would then check out a different version of the release test (possibly non-existing)

The `build_pipeline.py` checks out both the ray test repo/branch and the ray wheels repo/branch to find the respective latest commit. This happens in the course of a few seconds. So yes, this can be outdated, but it's much closer in time than what we currently have in different jobs. 

Generally I think it this way: The build_pipeline step should take full control over what is being run in the subsequent job. It resolves the wheels we want to test on, and it resolved the test repo+branch we want to run on. Moving this resolving to the run_release_test.sh makes it a bit complicated.

We have:

```
  git clone --depth 1 -b "${RAY_TEST_BRANCH}" "${RAY_TEST_REPO}" "${TMPDIR}"
```

in that script, I think checking out a possibly set `RAY_TEST_COMMIT` is a clean solution for the script and can be easily set by the pipeline builder. What do you think?

Reply:
This PR https://github.com/ray-project/ray/pull/34400/files introduces this RAY_TEST_COMMIT flag

Reply:
ah actually just saw this message, let me take a look.

Reply:
great! I am gonna use this interface then!

Reply:
Actually do you want to just include this into your PR then? I'm fine with that!